File: plugins/capacitor-metronome-background/android/src/main/java/com/zhengyaoliu/metrognome/background/MetronomeBackgroundPlugin.java
package com.zhengyaoliu.metrognome.background;

import android.content.Context;
import android.content.Intent;

import androidx.core.content.ContextCompat;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "MetronomeBackground", events = { "beat" })
public class MetronomeBackgroundPlugin extends Plugin {

    @Override
    public void load() {
        super.load();
        MetronomeForegroundService.setBeatListener(beatIndex -> {
            JSObject payload = new JSObject();
            payload.put("beatIndex", beatIndex);
            notifyListeners("beat", payload);
        });
    }

    @Override
    protected void handleOnDestroy() {
        MetronomeForegroundService.clearBeatListener();
        super.handleOnDestroy();
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        JSObject result = new JSObject();
        result.put("available", true);
        call.resolve(result);
    }

    @PluginMethod
    public void isAvailable(PluginCall call) {
        JSObject result = new JSObject();
        result.put("available", true);
        call.resolve(result);
    }

    @PluginMethod
    public void startPlayback(PluginCall call) {
        try {
            Context context = getContext().getApplicationContext();
            Intent intent = buildServiceIntent(context, MetronomeForegroundService.ACTION_START, call);
            ContextCompat.startForegroundService(context, intent);
            call.resolve();
        } catch (IllegalArgumentException error) {
            call.reject(error.getMessage());
        } catch (Exception error) {
            call.reject("Failed to start native metronome playback", error);
        }
    }

    @PluginMethod
    public void updatePlayback(PluginCall call) {
        try {
            Context context = getContext().getApplicationContext();
            Intent intent = buildServiceIntent(context, MetronomeForegroundService.ACTION_UPDATE, call);
            context.startService(intent);
            call.resolve();
        } catch (IllegalArgumentException error) {
            call.reject(error.getMessage());
        } catch (Exception error) {
            call.reject("Failed to update native metronome playback", error);
        }
    }

    @PluginMethod
    public void stopPlayback(PluginCall call) {
        try {
            Context context = getContext().getApplicationContext();
            Intent intent = new Intent(context, MetronomeForegroundService.class);
            intent.setAction(MetronomeForegroundService.ACTION_STOP);
            context.startService(intent);
            call.resolve();
        } catch (Exception error) {
            call.reject("Failed to stop native metronome playback", error);
        }
    }

    private Intent buildServiceIntent(Context context, String action, PluginCall call) {
        int bpm = call.getInt("bpm", 60);
        int beats = call.getInt("beats", 4);
        boolean stressFirst = call.getBoolean("stressFirst", true);
        String waveform = call.getString("waveform", "square");
        int[] subdivision = parseSubdivision(call.getArray("subdivision"));

        if (bpm < 1 || bpm > 400) {
            throw new IllegalArgumentException("bpm must be within 1..400");
        }
        if (beats < 1 || beats > 32) {
            throw new IllegalArgumentException("beats must be within 1..32");
        }

        Intent intent = new Intent(context, MetronomeForegroundService.class);
        intent.setAction(action);
        intent.putExtra(MetronomeForegroundService.EXTRA_BPM, bpm);
        intent.putExtra(MetronomeForegroundService.EXTRA_BEATS, beats);
        intent.putExtra(MetronomeForegroundService.EXTRA_STRESS_FIRST, stressFirst);
        intent.putExtra(MetronomeForegroundService.EXTRA_SUBDIVISION, subdivision);
        intent.putExtra(MetronomeForegroundService.EXTRA_WAVEFORM, waveform);
        return intent;
    }

    private int[] parseSubdivision(JSArray array) {
        if (array == null || array.length() == 0) {
            return new int[] { 1 };
        }

        int[] parsed = new int[array.length()];
        for (int index = 0; index < array.length(); index++) {
            int note = array.optInt(index, 1);
            parsed[index] = note == 0 ? 0 : 1;
        }
        return parsed;
    }
}

File: plugins/capacitor-metronome-background/ios/Sources/MetronomeBackgroundPlugin/MetronomeBackgroundPlugin.swift
import Capacitor
import Foundation

@objc(MetronomeBackgroundPlugin)
public class MetronomeBackgroundPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MetronomeBackgroundPlugin"
    public let jsName = "MetronomeBackground"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startPlayback", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "updatePlayback", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopPlayback", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "addBeatListener", returnType: CAPPluginReturnCallback),
        CAPPluginMethod(name: "removeBeatListener", returnType: CAPPluginReturnPromise),
    ]

    private let engine = MetronomeAudioEngine()
    private var savedBeatCall: CAPPluginCall?

    override public func load() {
        super.load()
        engine.onBeat = { [weak self] beatIndex in
            DispatchQueue.main.async {
                self?.savedBeatCall?.resolve(["beatIndex": beatIndex])
            }
        }
    }

    @objc func addBeatListener(_ call: CAPPluginCall) {
        if let previous = savedBeatCall {
            bridge?.releaseCall(previous)
        }
        call.keepAlive = true
        savedBeatCall = call
    }

    @objc func removeBeatListener(_ call: CAPPluginCall) {
        if let previous = savedBeatCall {
            bridge?.releaseCall(previous)
            savedBeatCall = nil
        }
        call.resolve()
    }

    @objc func initialize(_ call: CAPPluginCall) {
        call.resolve(["available": true])
    }

    @objc func isAvailable(_ call: CAPPluginCall) {
        call.resolve(["available": true])
    }

    @objc func startPlayback(_ call: CAPPluginCall) {
        runWithConfig(call, operationLabel: "start") { try engine.start(config: $0) }
    }

    @objc func updatePlayback(_ call: CAPPluginCall) {
        runWithConfig(call, operationLabel: "update") { engine.update(config: $0) }
    }

    @objc func stopPlayback(_ call: CAPPluginCall) {
        engine.stop()
        call.resolve()
    }

    // MARK: - Helpers

    private struct ConfigError: Error { let message: String }

    private func runWithConfig(
        _ call: CAPPluginCall,
        operationLabel: String,
        action: (MetronomeAudioEngine.Config) throws -> Void
    ) {
        do {
            let config = try parseConfig(call: call)
            try action(config)
            call.resolve()
        } catch let error as ConfigError {
            call.reject(error.message)
        } catch {
            call.reject("Failed to \(operationLabel) native metronome playback: \(error.localizedDescription)")
        }
    }

    private func parseConfig(call: CAPPluginCall) throws -> MetronomeAudioEngine.Config {
        let bpm = call.getInt("bpm") ?? 60
        let beats = call.getInt("beats") ?? 4
        let stressFirst = call.getBool("stressFirst") ?? true
        let waveform = MetronomeAudioEngine.normalizeWaveform(call.getString("waveform") ?? "square")

        guard (1...400).contains(bpm) else {
            throw ConfigError(message: "bpm must be within 1..400")
        }
        guard (1...32).contains(beats) else {
            throw ConfigError(message: "beats must be within 1..32")
        }

        var subdivision: [Int] = [1]
        if let raw = call.getArray("subdivision"), !raw.isEmpty {
            var parsed: [Int] = []
            parsed.reserveCapacity(raw.count)
            for item in raw {
                if let n = item as? Int {
                    parsed.append(n == 0 ? 0 : 1)
                } else if let n = item as? NSNumber {
                    parsed.append(n.intValue == 0 ? 0 : 1)
                } else {
                    parsed.append(1)
                }
            }
            if !parsed.isEmpty { subdivision = parsed }
        }

        return MetronomeAudioEngine.Config(
            bpm: bpm,
            beats: beats,
            stressFirst: stressFirst,
            subdivision: subdivision,
            waveform: waveform
        )
    }
}
